app/services/projects/archive_service.rb

```diff
@@ -48,7 +48,8 @@ def persist(service_call)
 
     def archive_project(project)
       # We do not care for validations but want the timestamps to be updated
-      project.update_attribute(:active, false)
+      # for the changed models.
+      project.update_attribute(:active, false) if project.active?
```

Comment:
Not sure whether these changes are required i.e, avoid updating the timestamp if the models are already archived. 

Note: These changes are not required to fix the issue but I've added this in case we don't want to update the timestamp of the models by archiving already archived projects.

Alternatively, we can only call `archive_project` for the subprojects which are active (by changing on line 43 of this file.)

Reply:
Hey @DubeySandeep, I think it is a good idea to change line 42 to only iterate over active projects and not change this method here.

And while you are already at line 42, do you mind to change it to iterate over all active subprojects and not only direct children (maybe you also meant that?). I think that would make it consistent with the permission checks in the contract which also check for subprojects and not only children. However, if you decide to change that please also cover that new behavior in the spec archiving a project hierarchy with a depth of at least 3.

Reply:
> I think it is a good idea to change line 42 to only iterate over active projects and not change this method here.

Thanks for the suggestion, I've updated the changes accordingly.

>  do you mind changing it to iterate over all active subprojects and not only direct children (maybe you also meant that?)

Yes, I've updated it to cover all the subprojects. (I had found this issue "only archiving next children" while testing.)

> I think that would make it consistent with the permission checks in the contract which also check for subprojects and not only children. However, if you decide to change that please also cover that new behavior in the spec archiving a project hierarchy with a depth of at least 3.


I agree, though I'm worried there might exist some subprojects in the datastore that users have assumed to be archived but they are still active. Also, I don't think we can run a one-off job to archive all such subprojects as (assuming) a subproject can be manually unarchived by a user, right? (I don't have much context here so leaving my thoughts in case it helps decide whether we need to take any actions for such subprojects.)


Also, I've added the test for the new functionality.